### PR TITLE
CM 1057 cleaner circleci and docker caching

### PR DESCRIPTION
**Pull request by rodriguesk**
Added support for:
- cleaner logs for pytest (separate section in circle ci report)
- docker caching (DLC = docker layer caching)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -17,25 +17,31 @@ jobs:
           name: Linter
           command: python3 -m black --check ./
   pytest:
-    machine: true
+    machine:
+      image: ubuntu-2004:202201-02
+      docker_layer_caching: true
```
Comment (danmash) [inline .circleci/config.yml:22]: This one works great! Major improvement!

Screenshot 2022-06-25 at 17 04 22

File: .circleci/config.yml [hunk 2]
```diff
@@ -17,25 +17,31 @@ jobs:
...
-          command: docker build -f docker/services/web/Dockerfile -t $DOCKER_REGISTRY_NAME/climatemindbackend:$CIRCLE_BUILD_NUM .
+          name: "Docker Compose Up"
+          command: docker-compose --file docker/docker-compose.pytest.yml up -d
+      - run:
+          name: "Wait for database to spin up"
+          command: sleep 30
+      - run:
+          name: "Run Pytest"
+          command: docker exec -it docker-web-1 pytest --cipdb
+          # or to NOT check for breakpoints accidentally left in the code
```
Comment (danmash) [inline .circleci/config.yml:38]: Don't see a reason for this comment

File: .circleci/config.yml [hunk 3]
```diff
@@ -17,25 +17,31 @@ jobs:
...
+          # docker-compose exec -it docker-web-1 pytest -n auto
```
Comment (danmash) [inline .circleci/config.yml:39]: wrong command, `-n auto` is for a parallel run which MSSQL doesn't support.

File: .circleci/config.yml [hunk 4]
```diff
@@ -17,25 +17,31 @@ jobs:
...
   cypress:
     # Set up test environment and run cypress tests
     machine:
       image: ubuntu-2004:202201-02
+      docker_layer_caching: true
```
Comment (danmash) [inline .circleci/config.yml:44]: This doesn't work for cypress for some reason, it was faster before. So I suggest to fix or remove

File: docker/services/web/pytest/entrypoint.pytest.sh [hunk 5]
```diff
@@ -3,8 +3,11 @@
...
+# python
+flask run --host=0.0.0.0
+
 # FIXME: unable to run pytest in parallel with xdist plugin due to the error
 #  Transaction (Process ID 52) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction. (1205) (SQLExecDirectW)')
 # pytest -n auto --cipdb
 
-pytest --cipdb
\ No newline at end of file
+# pytest --cipdb
```
Comment (danmash) [inline docker/services/web/pytest/entrypoint.pytest.sh:13]: commenting on everything is a bad practice. Just remove all comments if no longer needed

File: docker/services/web/pytest/entrypoint.pytest.sh [hunk 6]
```diff
@@ -3,8 +3,11 @@
 echo "sleeping entrypoint"
 sleep 120
 
+# python
+flask run --host=0.0.0.0
```
Comment (danmash) [inline docker/services/web/pytest/entrypoint.pytest.sh:7]: to be honest, we don't need any command in this container at all, especially our web app
so this doesn't make sense. 
I suggest to remove entrypoint script at all 
and use `sleep infinity` instead

add the following line here `docker/docker-compose.pytest.yml` to the `web` service.
 
    command: sleep infinity

**Review by danmash — CHANGES_REQUESTED** [review]
Do not merge without changes! And squash

File: .circleci/config.yml (hunk 4 above)
Comment (rodriguesk) [reply .circleci/config.yml:44]: I think there was just a weird issue with Cypress that day. It was sitting there for a while doing nothing as the clock was ticking up.

File: docker/services/web/pytest/entrypoint.pytest.sh (hunk 6 above)
Comment (rodriguesk) [reply docker/services/web/pytest/entrypoint.pytest.sh:7]: Yes sleep infinity should work too.

File: .circleci/config.yml (hunk 3 above)
Comment (rodriguesk) [reply .circleci/config.yml:39]: TSQL and the MSSQL server should support some degree of parallelism though. It should allow multiple reads from what I remember but only 1 write at a time. I think the server should manage it though and make a queue. But given how fast it currently is (15sec) I don’t think parallelizing is even needed.

File: .circleci/config.yml (hunk 2 above)
Comment (rodriguesk) [reply .circleci/config.yml:38]: To make sure that’s actually what’s happening. I wanted to ask you about that. Does it do that?

File: .circleci/config.yml (hunk 1 above)
Comment (danmash) [reply .circleci/config.yml:22]: @rodriguesk don't need to remove it, it works fine for the `pytest` step

**Pull request merged**